README.markdown
# Raca

A simple gem for interacting with Rackspace cloud APIs. The following APIs are
supported:

* Identity
* Cloud Files
* Cloud Servers

Raca intentionally has no dependencies outside the ruby standard library.

If loaded alongside Rails, it will utilise the rails cache to avoid repeated
requests to the rackspace identity API.

## Installation

    gem install raca

## Usage

For full usage details check the documentation for each class, but here's
a taste of the basics.

### Regions

Many of the Rackspace cloud products are available in multiple regions. When
required, you can specify a region using a symbol with the 3-letter region code.

Currently, the following regions are valid:

* :ord - Chicago
* :iad - Northern Virginia
* :syd - Sydney
* :dfw - Dallas-Fort Worth
* :hkg - Hong Kong

### Identity

To authenticate and begin any interaction with rackspace, you must create a
Raca::Account instance.

    account = Raca::Account.new("username", "api_key")

You can view the token that will be used for subsequent requests:

    puts account.auth_token

List the available APIs:

    puts account.service_names

... and then view the URLs for each service:

    puts account.public_endpoint("cloudFiles", :ord)

### Cloud Files

Using an existing Raca::Account object, retrieve a collection of Cloud Files
containers in a region like so:

    ord_containers = account.containers(:ord)

You can retrieve a single container from the collection:

    dir = ord_containers.get("container_name")

Retrieve some metadata on the collection:

    put ord_containers.metadata

With a single container, you can perform a range of operations on the container
and objects inside it.

    dir = ord_containers.get("container_name")

Download a file:

    dir.download("remote_key.txt", "/home/jh/local_file.txt")

Upload a file:

    dir.upload("target_path.txt", "/home/jh/local_file.txt")

List keys in the container, optionally limiting the results to those
starting with a prefix:

    puts dir.list
    puts dir.list(prefix: "subdir/")

Delete an object:

    dir.delete("target_path.txt")

View metadata on the container:

    puts dir.metadata
    puts dir.cdn_metadata

Enable access to the container contents via a public CDN. Use this with caution, it will make *all* objects public!

It accepts an argument telling the CDN edge nodes how long they can cache each object for (in seconds).

    dir.cdn_enable(60 * 60 * 24) # 1 day

Purge an object from the CDN:

    dir.purge_from_akamai("target_path.txt", "notify@example.com")

Generate a public URL to an object in a private container. The second argument
is the temp URL key that can be set using Raca::Containers#set_temp_url_key

    ord_containers = account.containers(:ord)
    ord_containers.set_temp_url_key("secret")
    dir = ord_containers.get("container_name")
    puts dir.temp_url("remote_key.txt", "secret", Time.now.to_i + 60)

### Cloud Servers

Using an existing Raca::Account object, retrieve a collection of Cloud Servers
from a region like so:

    ord_servers = account.servers(:ord)

You can retrieve a existing server from the collection:

    a_server = ord_servers.get("server_name")

Retrieve some details on the server:

    put a_server.metadata

You can use the collection to create a brand new server:

    a_server = ord_servers.create("server_name", "1Gb", "Ubuntu 10.04 LTS")

### Users

Using an existing Raca::Account object, retrieve a collection of Users like so:

    users = account.users

You can retrieve an existing user from the collection:

    a_user = users.get("username")

You can display details for each user with the details method:

    a_user = users.get("username")
    a_user.details

## General API principles

Methods that make calls to an API should never return a raw HTTP response
object. If a sensible return value is expected (retrieving metadata, listing
matches, etc) then that should always be returned. If return value isn't obvious
(change remote state, deleting an object, etc) then a simple boolean or similar
should be returned to indicate success.

If an unexpected error occurs (a network timeout, a 500, etc) then an exception
should be raised.


## Why not fog?

[fog](http://rubygems.org/gems/fog) is the [official](http://developer.rackspace.com)
ruby library for interacting with the Rackspace API. It is a very capable
library and supports much more of the API than this modest gem.

As of version 1.20.0, fog supports dozens of providers, contains ~152000 lines
of ruby and adds ~500ms to the boot time of our rails apps. raca is a
lightweight, rackspace-only alternative with minimal dependencies that should
have a negligable impact on application boot times. Version 0.3 has ~700 lines of
ruby (excluding specs). It also does *much* less than fog. We can't have our cake
and eat it too.

## Compatibility

The Raca version number is < 1.0 because it's highly unstable. Until we release
a 1.0.0, consider the API of this gem to be unstable.

## License

This library is released undr the MIT License. See the included MIT-LICENSE file
for further details


CHANGELOG
v0.4.4 (XXX)
* Add extra detail to exception messages raised when a HTTP request fails
* Increase read timeout on HTTP connections to 120s

v0.4.3 (12th September 2015)
* Use less memory when uploading large files (> 5Gb) to cloud files
* Allow a pre-computed ETag header to be provided when uploading to cloud files

v0.4.2 (19th July 2015)
* Large files (> 5Gb) that are uploaded to cloud files are broken into 4Gb
  chunks instead of 100Mb.

v0.4.1 (8th October 2014)
* Raca::Servers#create looks for an exact image name before falling back
  to a fuzzy match to allow precise selection of images.

v0.4.0 (25th August 2014)
* Added Raca::Container#temp_upload_url
* Renamed Raca::Container#expiring_url to Raca::Container#temp_url
  * the old method still exists for now but is deprecated
* Expand Raca::Container#metadata to include custom metadata
* Added raca::Container#set_metadata
* For non regioned APIs (like Cloud DNS) the region argument to Raca::Account#public_endpoint
  can be left off
* Added read access to user accounts via Raca::Account#users

v0.3.3 (26th April 2014)
* Added a User-Agent header to all requests
* Remove automatic retries after a rackspace timeout
* Bugfix: correctly upload files > 5Gb
* Bugfix: Use the correct ETag header for uploads

v0.3.2 (23rd April 2014)
* Bugfix: correctly handle a 401 response when a stale token is used for
  a streaming PUT

v0.3.1(14th April 2014)
* Add the rackspace transaction ID to HTTP exceptions

v0.3.0 (12th April 2014)
* Added an optional headers param to Raca::Container#upload
* Added the details option to Raca::Container#list
* Fixed Raca::Container#list to return all items for containers that
  contain more than 10,000 objects
* Fixed Raca::Container to work with container and object names that
  need to be escaped (spaces, some punctuation, utf8 characters, etc)
* Refactored internal management of HTTP requests to reduce duplication
  and improve testability. There should be no visible change to the public
  API of raca

v0.2.0 (19th March 2014)
* Breaking API changes to Raca::Servers and Raca:Server
  * Moved the create() method from Server to Servers
* Added Raca::Account#service_names
* Stop returning HTTP response objects from many methods on Raca::Container
* Added some custom error classes for common issues (timeouts, 404s, etc) to
  make recovering easier for the user

v0.1.1 (25th February 2014)
* Add Raca::Container#object_metadata

v0.1.0 (24th February 2014)
* Initial Release


lib/raca/account.rb
require 'yaml'
require 'json'

module Raca

  # This is your entrypoint to the rackspace API. Start by creating a
  # Raca::Account object and then use the instance method to access each of
  # the supported rackspace APIs.
  #
  class Account

    def initialize(username, key, cache = nil)
      @username, @key, @cache = username, key, cache
      @cache ||= if defined?(Rails)
        Rails.cache
      else
        {}
      end
    end

    # Return the temporary token that should be used when making further API
    # requests.
    #
    #     account = Raca::Account.new("username", "secret")
    #     puts account.auth_token
    #
    def auth_token
      extract_value(identity_data, "access", "token", "id")
    end

    # Return the public API URL for a particular rackspace service.
    #
    # Use Account#service_names to see a list of valid service_name's for this.
    #
    # Check the project README for an updated list of the available regions.
    #
    #     account = Raca::Account.new("username", "secret")
    #     puts account.public_endpoint("cloudServers", :syd)
    #
    # Some service APIs are not regioned. In those cases, the region code can be
    # left off:
    #
    #     account = Raca::Account.new("username", "secret")
    #     puts account.public_endpoint("cloudDNS")
    #
    def public_endpoint(service_name, region = nil)
      endpoints = service_endpoints(service_name)
      if endpoints.size > 1 && region
        region = region.to_s.upcase
        endpoints = endpoints.select { |e| e["region"] == region } || {}
      elsif endpoints.size > 1 && region.nil?
        raise ArgumentError, "The requested service exists in multiple regions, please specify a region code"
      end

      if endpoints.size == 0
        raise ArgumentError, "No matching services found"
      else
        endpoints.first["publicURL"]
      end
    end

    # Return the names of the available services. As rackspace add new services and
    # APIs they should appear here.
    #
    # Any name returned from here can be passe to #public_endpoint to get the API
    # endpoint for that service
    #
    #     account = Raca::Account.new("username", "secret")
    #     puts account.service_names
    #
    def service_names
      catalog = extract_value(identity_data, "access", "serviceCatalog") || {}
      catalog.map { |service|
        service["name"]
      }
    end

    # Return a Raca::Containers object for a region. Use this to interact with the
    # cloud files service.
    #
    #     account = Raca::Account.new("username", "secret")
    #     puts account.containers(:ord)
    #
    def containers(region)
      Raca::Containers.new(self, region)
    end

    # Return a Raca::Servers object for a region. Use this to interact with the
    # next gen cloud servers service.
    #
    #     account = Raca::Account.new("username", "secret")
    #     puts account.servers(:ord)
    #
    def servers(region)
      Raca::Servers.new(self, region)
    end

    # Raca classes use this method to occasionally re-authenticate with the rackspace
    # servers. You can probably ignore it.
    #
    def refresh_cache
      # Raca::HttpClient depends on Raca::Account, so we intentionally don't use it here
      # to avoid a circular dependency
      Net::HTTP.new('identity.api.rackspacecloud.com', 443).tap {|http|
        http.use_ssl = true
      }.start {|http|
        payload = {
          auth: {
            'RAX-KSKEY:apiKeyCredentials' => {
              username: @username,
              apiKey: @key
            }
          }
        }
        response = http.post(
          '/v2.0/tokens',
          JSON.dump(payload),
          {'Content-Type' => 'application/json'},
        )
        if response.is_a?(Net::HTTPSuccess)
          cache_write(cache_key, JSON.load(response.body))
        else
          raise_on_error(response)
        end
      }
    end

    # Return a Raca::HttpClient suitable for making requests to hostname.
    #
    def http_client(hostname)
      Raca::HttpClient.new(self, hostname)
    end

    def inspect
      "#<Raca::Account:#{__id__} username=#{@username}>"
    end

    private

    def raise_on_error(response)
      error_klass = case response.code.to_i
      when 400 then BadRequestError
      when 401 then UnauthorizedError
      when 404 then NotFoundError
      when 500 then ServerError
      else
        HTTPError
      end
      raise error_klass, "Rackspace returned HTTP status #{response.code}"
    end

    # This method is opaque, but it was the best I could come up with using just
    # the standard library. Sorry.
    #
    # Use this to safely extract values from nested hashes:
    #
    #     data = {a: {b: {c: 1}}}
    #     extract_value(data, :a, :b, :c)
    #     => 1
    #
    #     extract_value(data, :a, :b, :d)
    #     => nil
    #
    #     extract_value(data, :d)
    #     => nil
    #
    def extract_value(data, *keys)
      if keys.empty?
        data
      elsif data.respond_to?(:[]) && data[keys.first]
        extract_value(data[keys.first], *keys.slice(1,100))
      else
        nil
      end
    end

    # An array of all the endpoints for a particular service (like cloud files,
    # cloud servers, dns, etc)
    #
    def service_endpoints(service_name)
      catalog = extract_value(identity_data, "access", "serviceCatalog") || {}
      service = catalog.detect { |s| s["name"] == service_name } || {}
      service["endpoints"] || []
    end

    def cache_read(key)
      if @cache.respond_to?(:read) # rails cache
        @cache.read(key)
      else
        @cache[key]
      end
    end

    def cache_write(key, value)
      if @cache.respond_to?(:write) # rails cache
        @cache.write(key, value)
      else
        @cache[key] = value
      end
    end

    def identity_data
      refresh_cache unless cache_read(cache_key)

      cache_read(cache_key) || {}
    end

    def cache_key
      "raca-#{@username}"
    end

  end
end
